portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java

```diff
@@ -1341,6 +1341,7 @@ public User addUserWithWorkflow(
 
 		workflowServiceContext.setAttribute("autoPassword", autoPassword);
 		workflowServiceContext.setAttribute("sendEmail", sendEmail);
+		workflowServiceContext.setCompanyId(user.getCompanyId());
 
 		user = WorkflowHandlerRegistryUtil.startWorkflowInstance(
```

Comment:
@istvansajtos, I think you consider the case when serviceContext = null in the UserLocalServiceImpl#addUserWithWorkflow(). But the way does not cover the all cases. updateIncompleteUser() method also invoked this. I think we need to place the logic to the common place. 

Refer to WorkflowHandlerRegistryUtil.startWorkflowInstance(), the map workflowContext always include WorkflowConstants.CONTEXT_COMPANY_ID. So we may set this in UserWorkflowHandler#updateStatus().
```
if (serviceContext.getCompanyId() == 0) {
			serviceContext.setCompanyId(
				GetterUtil.getLong(workflowContext.get(
					WorkflowConstants.CONTEXT_COMPANY_ID)));
		}
```

Please also add serviceContext = null test case to cover the fix. You may extract the [codes](https://github.com/liferay-core-infra/liferay-portal/blob/b4cb2aad52beeec74497b0740def8e4ce7319423/modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java#L137-L145) to the private method, named _testAddUser(..., ServiceContext serviceContext), and then use serviceContext = null and serviceContext != null to create two users.